feat: Track when RPC update from network banner is completed

cryptodev-2s opened the pull request:
## **Description**

### Track RPC Update Completion from Network Banner

### Summary

Adds `NetworkConnectionBannerRpcUpdated` event to track when users **complete** the RPC update flow initiated from the network connection banner.

### Changes

- **New MetaMetrics event**: `NetworkConnectionBannerRpcUpdated`
- **Tracked on**: Network form submission when `trackRpcUpdateFromBanner` flag is set
- **Properties**:
  - `chain_id_caip`: CAIP-2 format (e.g., `eip155:1`)
  - `from_rpc_domain`: Original RPC domain being switched from (hostname for public RPCs, `'custom'` for private)
  - `to_rpc_domain`: New RPC domain being switched to (hostname for public RPCs, `'custom'` for private)

### Why

Currently we track when users **click** "Update RPC" (`NetworkConnectionBannerUpdateRpcClicked`), but not when they **complete** the update.

```mermaid
sequenceDiagram
    participant User
    participant Banner as Network Banner
    participant Form as Network Form
    participant Analytics

    Note over Banner: Network issue detected
    Banner->>User: Show "degraded" or "unavailable" banner
    Note over Analytics: 📊 NetworkConnectionBannerShown

    User->>Banner: Click "Update RPC"
    Banner->>Analytics: 📊 NetworkConnectionBannerUpdateRpcClicked
    Note right of Analytics: Existing event<br/>Tracks user INTENT

    Banner->>Form: Navigate to network settings
    User->>Form: Edit RPC endpoint
    User->>Form: Click "Save"

    Form->>Analytics: 📊 NetworkConnectionBannerRpcUpdated
    Note right of Analytics: NEW event (this PR)<br/>Tracks COMPLETION

    Form->>User: Network updated ✅
```

### Related

- Segment schema PR: https://github.com/Consensys/segment-schema/pull/366, https://github.com/Consensys/segment-schema/pull/381

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37751?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: null

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/WPC-172

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds `NetworkConnectionBannerRpcUpdated` event and wires a banner->form flag to track completed RPC updates with CAIP chain ID and RPC domain metadata.
> 
> - **Analytics (MetaMetrics)**:
>   - Add `MetaMetricsEventName.NetworkConnectionBannerRpcUpdated`.
> - **State & Actions**:
>   - Extend `setEditedNetwork`/`editedNetwork` with `trackRpcUpdateFromBanner`.
>   - Update `getEditedNetwork` selector types accordingly.
> - **UI Flow**:
>   - `network-connection-banner`: on "Update RPC", dispatch `setEditedNetwork({ chainId, trackRpcUpdateFromBanner: true })` and navigate to `NETWORKS_ROUTE`.
>   - `network-list-menu`: pass `trackRpcUpdateFromBanner` to `NetworksForm`.
>   - `NetworksForm`: on Save (when editing) and flag is set, track `Network Connection Banner RPC Updated` with:
>     - `chain_id_caip` (`eip155:<decimal>`),
>     - `from_rpc_domain`/`to_rpc_domain` (public host or `custom`/`unknown`).
>   - Utilize `hexToNumber`, `isPublicEndpointUrl`, `onlyKeepHost` for payload.
> - **Tests**:
>   - Update banner tests to assert the new flag in `setEditedNetwork`.
>   - Add form tests covering event emission, absence when flag not set, custom endpoints, and missing `rpcEndpoints` handling.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 07049764571edabca83d8a715d12e69bf49ea92a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

cryptodev-2s committed:
metrics: Track when RPC update from network banner is completed

Add NetworkConnectionBannerRpcUpdated event to track when users
complete the RPC update flow initiated from the connection banner.

Includes chain_id_caip and rpc_endpoint_url properties.

mcmire commented on ui/pages/settings/networks-tab/networks-form/networks-form.test.js:
```diff
@@ -87,6 +88,12 @@ describe('NetworkForm Component', () => {
+    updateNetwork.mockClear(); // Reset calls from network update tests
+    addNetwork.mockClear(); // Reset calls from network creation tests
+    setTokenNetworkFilter.mockClear(); // Reset calls from token filter tests
+
```
If we call `jest.clearAllMocks()` instead, does that still work?

``` suggestion
    jest.clearAllMocks();
```

mcmire reviewed (commented):
A couple of comments.

cryptodev-2s commented on ui/pages/settings/networks-tab/networks-form/networks-form.tsx:
```diff
@@ -309,6 +312,35 @@ export const NetworksForm = ({
+                infuraProjectId ?? '',
+              )
+                ? onlyKeepHost(selectedRpcEndpoint.url)
+                : 'custom';
```
Yes, that makes perfect sense. I forgot to remove this, I was previously using rpcUrls directly, but not anymore.

cryptodev-2s committed:
refactor: Remove unnecessary undefined check in RPC tracking

selectedRpcEndpoint cannot be undefined here because:
- Save button is disabled without rpcEndpoints
- trackRpcUpdateFromBanner only set for existing networks
- Existing networks always have valid RPC endpoints

cryptodev-2s committed:
metrics: Track RPC migration path (from/to) in banner update event

Change from single rpc_endpoint_url to from_rpc_domain and to_rpc_domain
properties to track the complete migration path when users switch RPCs
from the network connection banner.

This enables analytics to identify:
- Which RPC endpoints are problematic (high from_rpc_domain counts)
- Which RPC endpoints users prefer (high to_rpc_domain counts)
- Common migration patterns (FROM → TO pairs)

cryptodev-2s committed:
fix: Handle corrupted state in RPC tracking gracefully

- Use optional chaining for existingNetwork.rpcEndpoints
- Return 'unknown' for missing/corrupted from_rpc_domain
- Add test for corrupted state scenario

This prevents crashes if existingNetwork lacks rpcEndpoints while
still capturing analytics data to identify data integrity issues.

cryptodev-2s commented on ui/pages/settings/networks-tab/networks-form/networks-form.tsx:
```diff
@@ -309,6 +312,40 @@ export const NetworksForm = ({
+                from_rpc_domain: oldRpcEndpoint?.url
+                  ? sanitizeRpcUrl(oldRpcEndpoint.url)
+                  : 'unknown',
+                to_rpc_domain: sanitizeRpcUrl(newRpcEndpoint.url),
```
Form validation guarantees `rpcUrl` exists in `rpcEndpoints` before submission, so `newRpcEndpoint` is always defined. The `defaultRpcEndpointIndex` is set via `findIndex` on the validated `rpcUrl`.

mcmire reviewed (approved):
Tested and LGTM.
